File: C++/DesignPatterns/FactoryMethodDesignPattern.cpp
#include <memory>
#include <stdexcept>
#include <string>

// product
struct Notification {
    virtual void send(const std::string& message) = 0;
    virtual ~Notification() = default;
};

struct SMSNotification : Notification {
    void send(const std::string& message) override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct SlackNotification : Notification {
    void send(const std::string& message) override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct WhatsappNotification : Notification {
    void send(const std::string& message) override {
        throw std::runtime_error("Unimplemented method 'send'");
    }
};

struct PushNotification : Notification {
    void send(const std::string& message) override {}
};

// product factory
struct NotificationFactory {
    virtual std::unique_ptr<Notification> getNotification() = 0;
    virtual ~NotificationFactory() = default;
};

struct SMSNotificationFactory : NotificationFactory {
    std::unique_ptr<Notification> getNotification() override {
        return std::make_unique<SMSNotification>();
    }
};

struct WhatsappNotificationFactory : NotificationFactory {
    std::unique_ptr<Notification> getNotification() override {
        return std::make_unique<WhatsappNotification>();
    }
};

struct PushNotificationFactory : NotificationFactory {
    std::unique_ptr<Notification> getNotification() override {
        return std::make_unique<PushNotification>();
    }
};

class Application {
public:
    void sendNotification(NotificationFactory& factory, const std::string& type) {
        auto notification = factory.getNotification();
        notification->send("message");
    }
};

// Factory Method Design Pattern

int main() {
    return 0;
}

File: Python/DesignPatterns/FactoryMethodDesignPattern.py
from abc import ABC, abstractmethod


# product
class Notification(ABC):
    @abstractmethod
    def send(self, message: str) -> None:
        raise NotImplementedError


class SMSNotification(Notification):
    def send(self, message: str) -> None:
        raise NotImplementedError("Unimplemented method 'send'")


class SlackNotification(Notification):
    def send(self, message: str) -> None:
        raise NotImplementedError("Unimplemented method 'send'")


class WhatsappNotification(Notification):
    def send(self, message: str) -> None:
        raise NotImplementedError("Unimplemented method 'send'")


class PushNotification(Notification):
    def send(self, message: str) -> None:
        pass


# product factory
class NotificationFactory(ABC):
    @abstractmethod
    def get_notification(self) -> Notification:
        raise NotImplementedError


class SMSNotificationFactory(NotificationFactory):
    def get_notification(self) -> Notification:
        return SMSNotification()


class WhatsappNotificationFactory(NotificationFactory):
    def get_notification(self) -> Notification:
        return WhatsappNotification()


class PushNotificationFactory(NotificationFactory):
    def get_notification(self) -> Notification:
        return PushNotification()


class Application:
    def send_notification(self, factory: NotificationFactory, message_type: str) -> None:
        notification = factory.get_notification()
        notification.send("message")


# Factory Method Design Pattern


def main() -> None:
    pass


if __name__ == "__main__":
    main()
